Instant Purchase Unit Tests (Partial Coverage)

pmclain opened the pull request:
### Description
Begins adding some test coverage for Magento_InstantPurchase. Minor refactoring was performed while writing the test cases.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Add Block Unit Tests

Commit:
Use correct interfaces in type hinting and phpdocs

Type hints and phpdocs should be the appropriate interface and not the
class defined by preference in `di.xml`

Commit:
Minor refactor of controller for testability

The purpose is to limit the repeated calls to `getParam`

Commit:
Add controller unit test

Commit:
Use StoreInferface instead of Store Model in hints

Commit:
Remove second call to isAvailable

Commit:
Add test for CustomerData/InstantPurchase

Commit:
Add config model test

Commit:
Add unit test for PlaceOrder model

Commit:
Add unit test for InstantPurchaseChooser model

Commit:
Unit test for InstantPurchaseOption model

Commit:
Unit test for DefaultBillingAddressChooser

vkublytskyi commented on app/code/Magento/InstantPurchase/Controller/Button/PlaceOrder.php:
```diff
@@ -104,21 +104,22 @@ public function __construct(
     {
-        $request = $this->getRequest();
-        if (!$this->doesRequestContainAllKnowParams($request)) {
+        $params = $this->getRequest()->getParams();
```
These changes decrease readability of code and introduce code duplication. Testability is not affected anyhow. Controller invokes only two methods of the request: `getParam` and `getParams`. Fetching all parameters into array and use it in procedural style is not helps with testability.

To mock request properly `PHPUnit_Framework_MockObject_Builder_InvocationMocker::willReturnMap` or `PHPUnit_Framework_MockObject_Builder_InvocationMocker::willReturnCallback` may be used.

vkublytskyi commented on app/code/Magento/InstantPurchase/Controller/Button/PlaceOrder.php:
```diff
@@ -104,21 +104,22 @@ public function __construct(
-        $productId = (int)$request->getParam('product');
-        $productRequest = $this->getRequestUnknownParams($request);
+        $paymentTokenPublicHash = (string)$params['instant_purchase_payment_token'];
+        $shippingAddressId = (int)$params['instant_purchase_shipping_address'];
```
Such refactoring decrease quality of code:
1. 6 rows now contains duplicated logic
2. Instead of reading simply "get named parameter from request" now developer need to remember that `$params` were fetched from a request and interpret ternary operator.

vkublytskyi commented on app/code/Magento/InstantPurchase/Controller/Button/PlaceOrder.php:
```diff
@@ -104,21 +104,22 @@ public function __construct(
+        $paymentTokenPublicHash = (string)$params['instant_purchase_payment_token'];
+        $shippingAddressId = (int)$params['instant_purchase_shipping_address'];
+        $billingAddressId = (int)$params['instant_purchase_billing_address'];
+        $carrierCode = (string)isset($params['instant_purchase_carrier']) ? $params['instant_purchase_carrier'] : '';
```
This line contains an error (and this is one of the reasons why working with request object are preferable over fetching data into an array and usage of `isset`): type casting will be applied to the condition of the ternary operator, not to the "true" part of result. Se  code example https://3v4l.org/viTNs

vkublytskyi commented on app/code/Magento/InstantPurchase/CustomerData/InstantPurchase.php:
```diff
@@ -94,7 +94,7 @@ public function getSectionData(): array
             'available' => $instantPurchaseOption->isAvailable()
         ];
-        if (!$instantPurchaseOption->isAvailable()) {
+        if (!$data['available']) {
```
This also decrease code readability as now code reader should keep in mind how `$data['available']` was retrieved. Previous code holds logic based on `$instantPurchaseOption` and fill the data (it does not make a decision based on a result, it produce a result from an input).

vkublytskyi commented on app/code/Magento/InstantPurchase/Model/InstantPurchaseChooser.php:
```diff
@@ -11,7 +11,7 @@
 use Magento\InstantPurchase\Model\BillingAddressChoose\BillingAddressChooserInterface;
 use Magento\InstantPurchase\Model\ShippingMethodChoose\ShippingMethodChooserInterface;
-use Magento\Store\Model\Store;
+use Magento\Store\Api\Data\StoreInterface;
```
Models (Store, Product, Quote) is not used in Instant Purchase by mistake. This was done by intent as models have more methods than API data interfaces. If data interfaces will be used instead of models then at some parts  Instant Purchase will call unknown methods which will work only if model is passed as data interface implementation.

Magento 2 indeed encourage usage of API data interfaces but programming based on contract have more importance. Instant Purchase will not work if someone substitutes data interface preference with own implementation which will not extend corresponding model.
Unfortunately, implementation of  Instant Purchase also relies on the current behavior of Quote module (e.g. save of quote in the middle of order placement at the `\Magento\InstantPurchase\Model\PlaceOrder::placeOrder`).

Thereby, dependency on models (which also marked as `@api`) is actually declaration on what code really depends.

Please revert back dependency on model classes.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Block/ButtonTest.php:
```diff
@@ -0,0 +1,143 @@
+     */
+    public function testIsEnabled($status)
+    {
+        $this->configMock->expects($this->once())
```
Please don't use `expects` in mocks as this, actually, not test a code but just "freeze" current implementation.

The only valid usage of `excpects` in unit tests are:
- you know that some code has side effects (which is bad practice)
- you fix a performance issue and would like to have a test that heavy method is not invoked more than required to make a job done.

In most cases, method call should return same result for same input arguments and simple method call is not a performance issue (avoid premature optimization). This comes to conclusion that PHPUnit mocks in most cases should mostly map output to input (it's actually not a mock but stub).

There is a great article by Uncle Bob on types of test doubles: https://8thlight.com/blog/uncle-bob/2014/05/14/TheLittleMocker.html.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Controller/Button/PlaceOrderTest.php:
```diff
@@ -0,0 +1,344 @@
+            ->method('setData')
+            ->with(['response' => self::GENERIC_ERROR_MESSAGE]);
+        
+        $this->messageManagerMock->expects($this->once())
```
Do not test that you call some particular method. With such code you freeze implementation. Instead, test that message manager has a message after  controller execution (this is more relevant to what you are expecting: you want that message is in messages list, you don't care how it appears there). Fake message manager may be used for this purpose.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Controller/Button/PlaceOrderTest.php:
```diff
@@ -0,0 +1,344 @@
+            ->method('addErrorMessage')
+            ->with(self::GENERIC_ERROR_MESSAGE);
+        
+        $this->assertInstanceOf(
```
In addition to result type verification, a content of the response should be checked as well. Having JSON result is not enough to be sure that response of controller is correct.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Controller/Button/PlaceOrderTest.php:
```diff
@@ -0,0 +1,344 @@
+            ->with(['response' => $successMessage]);
+
+        $this->messageManagerMock->expects($this->once())
+            ->method('addSuccessMessage')
```
Probably, we should not verify exact success message. The thing that we really want to verify is that success message *contains* order number.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/CustomerData/InstantPurchaseTest.php:
```diff
@@ -0,0 +1,143 @@
+            ->method('isAvailable')
+            ->willReturn(true);
+
+        $this->assertTrue($this->instantPurchase->getSectionData()['available']);
```
The *format* of customer data segment is an API on which depends client-side application. It change would be a backward incompatible. And this is actually a main risk during this class modification. Unit test should verify a *format* of output data.

Instant Purchase module have a good definition what is input of the method and what is output. Try to build most assertions on methods output and avoid use of mock test double (do not assert that some method is invoked). You may use fake objects and stubs to isolate tested class and provide some basic implementation.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Block/ButtonTest.php:
```diff
@@ -0,0 +1,143 @@
+
+    protected function setUp()
+    {
+        $objectManager = new ObjectManager($this);
```
Try to avoid usage of object manager in unit tests. Unit tests should be independent of infrastructure as much as possible.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Model/InstantPurchaseChooserTest.php:
```diff
@@ -0,0 +1,99 @@
+            ->method('isModuleEnabled')
+            ->willReturn(true);
+
+        $this->optionFactoryMock->expects($this->once())
```
This test does not verify that option object is created and fulfilled correctly.  It just verify that if IP enabled for store it will try to instantiate option. Please add assertions on the method result.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Model/InstantPurchaseOptionTest.php:
```diff
@@ -0,0 +1,186 @@
+     */
+    public function testIsAvailable($available)
+    {
+        $option = $this->objectManager->getObject(
```
You may use `new` operator in unit test. (You want to verify particular class, no need to manage objects or replace them with interceptors, right?)

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Model/InstantPurchaseOptionTest.php:
```diff
@@ -0,0 +1,186 @@
+            ]
+        );
+
+        $this->assertInstanceOf(Address::class, $option->getShippingAddress());
```
Dead code. `assertInstanceOf` will be never invoked as exception will be thrown before.
It is enough to have:
```php
/**
 * @expectedException \Magento\Framework\Exception\LocalizedException
 */
public function testGetShippingAddressException()
{
     $option = new InstantPurchaseOption(null, null, null, null);
     $option->getShippingAddress();
}
```

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Model/InstantPurchaseOptionTest.php:
```diff
@@ -0,0 +1,186 @@
+            ->method('getAvailable')
+            ->willReturn($available);
+
+        $this->assertEquals($available, $option->isAvailable());
```
This test does not verify correct availability logic (actually, only logic in the class) without addresses and payment token. Suppose data provider will be a good solution to implement such tests.

vkublytskyi commented on app/code/Magento/InstantPurchase/Test/Unit/Model/PlaceOrderTest.php:
```diff
@@ -0,0 +1,274 @@
+        );
+    }
+
+    public function testPlaceOrder()
```
This class is actually script for invoke other classes in exact order. It has no own logic and totally rely on current behavior of other modules. Unit test will not provide usefull verification here. Valid data and correct work verification may be provided only by integration tests here.
BTW. Fact that implementation have to relay of internal qoute and payment behavior was a reason to not introduce interface but implement order placement with a class.

pmclain replied:
Shouldn't there at least be a test confirming quotes are made inactive when catching exceptions? Failing to do so would replace a customer's current cart with the failed instant purchase cart on the frontend.

vkublytskyi replied:
Yes, fact that quote should be deactivated in case of failure should be verified. But it better to place in integration test as well (simple verification that `setIsActive` is invoked with `false` is not enough, what we really want to test is if order placement failed then after quote object load from repository it will be not active). For simplicity in such case (side effect of the code) we may verify that quote repository `save` method is called and quote object is not active already at this time.

Ideally, we should not create/save quote until order placement but this may lead to failures in the shipping configuration. So as this code is for integration it's better to test with integration test.

vkublytskyi reviewed (changes requested):
@pmclain Thank you for working on test coverage for Instant Purchase module. I have several high-level remarks according to your implementation:
1. Models are used at some places instead of data interfaces as implementation requires some methods that are defined only in models but not in interfaces and depends on current models behavior. Explicit dependency on modules declare this dependency. It is not good but ok as all used modules marked with an `@api` annotation
2. Instant Purchase uses slightly defined input and output for most of the methods. Unit tests should do assertions on output, not verify invocation of tested class collaborators (use fake and stubs test doubles instead of mocks).
3. Refactoring should increase code readability for humans. If some code is hard to test with PHPunit infrastructure the only possible refactoring is to extract part of logic to another class and cover it with another set of unit tests.  Rewriting code to reduce amount of methods that should be mocked is not a reasonable approach as it complicates implementation of class under test and make code less object oriented.

vkublytskyi commented:
@pmclain As I mentioned in review Instant Purchase code sometimes depends on the internal behavior of Quote and Payment modules. That was necessary to make it work. Unit tests are not good tool to verify such implementation so you may be also interested in extending integration tests coverage: https://github.com/magento/magento2/tree/2.2-develop/dev/tests/integration/testsuite/Magento/InstantPurchase

Commit:
Revert "Use correct interfaces in type hinting and phpdocs"

This reverts commit 28d9ff000057e7b219926fadc87a5ab69265ba6b.

Commit:
Revert "Minor refactor of controller for testability"

This reverts commit 1dabb2f2a6bfc65e0dd438623dc360f15efd14b8.

Commit:
Revert "Use StoreInferface instead of Store Model in hints"

This reverts commit 2566f4d2c77ca6054cf50dde2698d968f5a7b3bd.

Commit:
Revert "Remove second call to isAvailable"

This reverts commit 86df7f231b00d9a17dafddca06f58a253bac24fd.

Commit:
PlaceOrderTest review revisions

Commit:
Refactor InstantPurchaseOptionTest

Commit:
Revert "PlaceOrderTest review revisions"

This reverts commit b074e3f167026213c07697395c4347494ca7e3db.

Commit:
Remove InstantPurchaseOptionTest

Commit:
Remove InstantPurchaseChooserTest

No real value here, return object is the same regardless of logic. The
object contents will differ, but that is not best tested here.

Commit:
Test output format in InstantPurchaseTest

Commit:
Refactor ButtonTest

Commit:
Remove PlaceOrder controller test

pmclain commented:
@vkublytskyi Thanks for the guidance! I've updated the tests based on the feedback provided. The test for the place order controller has been removed, hopefully I'll find a time block to revisit it soon.
